resample/bootstrap.py

```diff
@@ -0,0 +1,285 @@
+"""
+Bootstrap resampling.
+"""
+from typing import Callable, Optional, Tuple, Sequence, Union
+import numpy as np
+from scipy import stats
+from resample.jackknife import jackknife as _jackknife
+from resample.empirical import quantile_gen as _quantile_gen
+
+
+def resample(
```

Comment:
I'm a little confused by this extra layer of abstraction, could we not wrap this logic into the bootstrap function itself? I haven't looked too closely but it seems potentially duplicative given that this is returning bootstrap samples. If we did have this it seems it might make more sense as a private helper function than a user-facing one.

Reply:
I thought we had discussed this before and agreed to use the `bootstrap` exclusively for "array of function applied to bootstrap replicas". `resample` is not a function but a generator of bootstrap replicas. This split allows one to get one replicated sample at a time instead to do with whatever you want, instead of generating them all at once. We cannot fold this into `bootstrap`, because `bootstrap` is a function. It would be weird to use the sample object as a function and as a generator.

Reply:
The stratification code looks simpler, now, too.

Reply:
And it is for the symmetry with jackknife (although that is the least important reason).

Here is another argument for adding the generator separately. When designing an interface, one should ask: what is the minimum amount of useful work that the interface should do for the user. In this case, the minimum amount of useful work is to generate a single replicated sample (either jackknife'd or bootstrapped), which the use can use in any way they want. Getting a single replicated sample at a time is better than an interface that precomputes all samples and only allows me to get them all at once. Because in the latter design, I have to hold all the replicas in memory at once, while in the former design I only need to hold on replica in memory.

In case of the parametric and ordinary bootstrap, there are large savings in memory consumption when you use a generator. Python loves generators for that reason, they replaced a lot of the former sequences with generators (e.g. `range`). They are more memory efficient than fully realized sequences.

Reply:
Ok, but I think the generator should be user-facing. We are exposing something useful. For example, if I want to apply my function in parallel on all the resampled samples, I would do something like this:
```py
from concurrent.futures import ProcessPoolExecutor as Pool
def myfn(x): ...
data = ...
with Pool() as p:
    for  replicas in p.map(myfn, resample(data)):
        # do stuff with replicas
```

Reply:
The resample generator is really the core element of this library, everything else is sugar and convenience on top.

Reply:
Actually..., I think I need to fix the jackknife generator so that returns copies by default, otherwise the above code would not work for the jackknife. The inplace modification should be opt-in to increase performance when possible.

Reply:
Yeah, below I convinced myself that it makes sense to expose it (and remembered we'd mentioned it before)